kyaml/setters2/settersutil/substitutioncreator.go

```diff
@@ -62,3 +74,138 @@ func (c SubstitutionCreator) Create(openAPIPath, resourcesPath string) error {
 		Outputs: []kio.Writer{inout},
 	}.Execute()
 }
+
+// CreateSettersForSubstitution creates the setters for all the references in the substitution
+// values if they don't already exist in openAPIPath file.
+func (c SubstitutionCreator) CreateSettersForSubstitution(openAPIPath string) error {
+	b, err := ioutil.ReadFile(openAPIPath)
+	if err != nil {
+		return err
+	}
+
+	// parse the yaml file
+	y, err := yaml.Parse(string(b))
+	if err != nil {
+		return err
+	}
+
+	m, err := c.GetValuesForMarkers()
+	if err != nil {
+		return err
+	}
+
+	// for each ref in values, check if the setter already exists, if not create them
+	for _, value := range c.Values {
+		obj, err := y.Pipe(yaml.Lookup(
+			// get the setter key from ref. Ex: from #/definitions/io.k8s.cli.setters.image_setter
+			// extract io.k8s.cli.setters.image_setter
+			"openAPI", "definitions", strings.TrimPrefix(value.Ref, "#/definitions/")))
+
+		if err != nil {
+			return err
+		}
+
+		if obj == nil {
+			sd := setters2.SetterDefinition{
+				// get the setter name from ref. Ex: from #/definitions/io.k8s.cli.setters.image_setter
+				// extract image_setter
+				Name:  strings.TrimPrefix(value.Ref, "#/definitions/io.k8s.cli.setters."),
+				Value: m[value.Marker],
+			}
+			err := sd.AddToFile(openAPIPath)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// GetValuesForMarkers parses the pattern and field value to derive values for the
+// markers in the pattern string. Returns error if the marker values can't be derived
+func (c SubstitutionCreator) GetValuesForMarkers() (map[string]string, error) {
+	m := make(map[string]string)
+	indices, err := c.GetStartIndices()
+	if err != nil {
+		return nil, err
+	}
+	s := c.FieldValue
+	p := c.Pattern
+	i := 0
+	j := 0
+	// iterate s, p with indices i, j respectively and when j hits the index of a marker, freeze j and iterate
+	// i and capture string till we find the substring just after current marker and before next marker
+
+	// Ex: s = "something/ubuntu:0.1.0", p = "something/IMAGE::VERSION", till j reaches 10
+	// just proceed i and j and check if s[i]==p[j]
+	// when j is 10, freeze j and move i till it sees substring '::' which derives IMAGE = ubuntu and so on.
+	for i < len(s) && j < len(p) {
+		if marker, ok := indices[j]; ok {
+			value := ""
+			e := j + len(marker)
+
+			for i < len(s) && (e == len(p) ||
+				s[i:min(len(s), i+lenToNextMarker(indices, e))] != p[e:min(e+lenToNextMarker(indices, e), len(p))]) {
+				value += string(s[i])
+				i++
+			}
+			// if marker is repeated in the pattern, make sure that the corresponding values
+			// are same and throw error if not.
+			if prevValue, ok := m[marker]; ok && prevValue != value {
+				return nil, errors.Errorf("Same marker is found to have different values in field value.")
+			}
+			m[marker] = value
+			j += len(marker)
```

Comment:
its pretty hard to follow these lines -- at a minimum these should have comments explain why this value is being incremented

WDYT of factoring whole thing into a struct with functions?